contracts/Compass.vy

```diff
@@ -173,44 +230,50 @@ def update_valset(consensus: Consensus, new_valset: Valset):
 # This makes calls to contracts that execute arbitrary logic
 # message_id is to prevent replay attack and every message_id can be used only once
 @external
-def submit_logic_call(consensus: Consensus, args: LogicCallArgs, message_id: uint256, deadline: uint256):
-    assert block.timestamp <= deadline, "Timeout"
+def submit_logic_call(consensus: Consensus, args: LogicCallArgs, fee_args: FeeArgs, message_id: uint256, deadline: uint256, gas_estimate: uint256):
+    self.gas_check(gas_estimate)
+    self.deadline_check(deadline)
     assert not self.message_id_used[message_id], "Used Message_ID"
     self.message_id_used[message_id] = True
     # check if the supplied current validator set matches the saved checkpoint
-    assert self.last_checkpoint == self.make_checkpoint(consensus.valset), "Incorrect Checkpoint"
+    self.check_checkpoint(self.make_checkpoint(consensus.valset))
     # signing data is keccak256 hash of abi_encoded logic_call(args, message_id, compass_id, deadline)
-    args_hash: bytes32 = keccak256(_abi_encode(args, message_id, compass_id, deadline, method_id=method_id("logic_call((address,bytes),uint256,bytes32,uint256)")))
+    args_hash: bytes32 = keccak256(_abi_encode(args, fee_args, message_id, compass_id, deadline, msg.sender, gas_estimate, method_id=method_id("logic_call((address,bytes),(uint256,uint256,bytes32),uint256,bytes32,uint256,address,uint256)")))
     # check if enough validators signed args_hash
     self.check_validator_signatures(consensus, args_hash)
     # make call to logic contract
     raw_call(args.logic_contract_address, args.payload)
+    FeeManager(FEE_MANAGER).transfer_fees(fee_args, unsafe_mul(tx.gasprice, gas_estimate), msg.sender)
```

Comment:
Thanks for clarifying. That leaves a small window of change though, as Paloma will have to ask "Does this user have sufficient funds to pay for the reimbursement". And obviously that will  depend on the gas cost of the block.

What's your idea on how to do this then?

Reply:
I think you can calculate it with approximate gasprice and gas_estimate and compare with user's funds. That means if it looks like the user has sufficient funds, the transaction might fail as insufficient funds. It is an edge case but we need to calculate gas cost with exact gasprice. Can this work?

Reply:
But at this stage the relayer will already have sent the transaction and gas costs will have incurred, right? That's why my original design worked like this:

- estimate cost of gas before transaction
- check user funds are sufficient to cover it
- send tx (and include the estimate in the payload)
- compass checks if user still able to cover the estimate, and then refunds the estimate (not the actual gas cost)

That means that sometimes it will overpay and sometimes underpay, but the relayer will **always** be reimbursed. Otherwise there might be calls where the relayer will not get reimbursement at all.

What do you think?

Reply:
Especially because all the fee amounts we also send in absolute units.

Reply:
We can check gas amount with remaining gas, not gas cost. And if the funds of user is insufficient because of gas price raising, the transaction will revert. So relayer must receive funds from successful SLC.